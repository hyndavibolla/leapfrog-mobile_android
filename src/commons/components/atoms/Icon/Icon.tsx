import React, { memo, useMemo } from 'react';
import { StyleProp, View, ViewStyle } from 'react-native';
import { createIconSetFromIcoMoon } from 'react-native-vector-icons';

import icoMoonConfig from './selection.json';
import { FONT_SIZE, COLOR, ICON } from '_constants';
import { useTestingHelper } from '_utils/useTestingHelper';

export interface Props {
  name: ICON;
  size?: FONT_SIZE;
  color?: COLOR;
  backgroundStyle?: StyleProp<ViewStyle>;
  innerBackgroundStyle?: StyleProp<ViewStyle>;
  [x: string]: unknown;
}

const CustomIcon = createIconSetFromIcoMoon(icoMoonConfig);

const Icon = ({ name, size = FONT_SIZE.REGULAR, color = COLOR.PRIMARY_BLUE, backgroundStyle, innerBackgroundStyle, ...rest }: Props) => {
  const { getTestIdProps } = useTestingHelper('icon');
  const parsedSize = Number.parseInt(size.replace('px', ''), 10);

  const customIcon = useMemo(
    () => <CustomIcon name={name} color={color} size={parsedSize} {...getTestIdProps('text')} {...rest} />,
    [color, getTestIdProps, name, parsedSize, rest]
  );

  // Normal Icon
  if (typeof backgroundStyle === 'undefined') return customIcon;

  // Icon with background
  return (
    <View
      style={[
        {
          borderRadius: parsedSize,
          width: parsedSize * 1.75,
          height: parsedSize * 1.75,
          justifyContent: 'center',
          alignItems: 'center'
        },
        backgroundStyle
      ]}
    >
      <View style={innerBackgroundStyle}>{customIcon}</View>
    </View>
  );
};

export default memo(Icon);
