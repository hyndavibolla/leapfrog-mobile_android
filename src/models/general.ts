import { Dispatch } from 'react';
import Auth0 from 'react-native-auth0';

import { ApiService } from '_services/ApiService';
import { AppService } from '_services/AppService';
import { CCPAService } from '_services/CCPAService';
import { ICrashlyticsService } from '_services/CrashlyticsService';
import { EventTrackerService } from '_services/EventTrackerService';
import { IHttpService } from '_services/HttpService';
import { NativeHelperService } from '_services/NativeHelperService';
import { Logger } from '_services/Logger';
import { IPerformanceTrackingService } from '_services/PerformanceTrackingService';
import { RaiseService } from '_services/RaiseService';
import { RemoteConfigService } from '_services/RemoteConfigService';
import { ICardLinkState } from '_state_mgmt/cardLink/state';
import { ICoreState } from '_state_mgmt/core/state';
import { IGameState } from '_state_mgmt/game/state';
import { IGiftCardState } from '_state_mgmt/giftCard/state';
import { IMessagesState } from '_state_mgmt/messages/state';
import { IMissionState } from '_state_mgmt/mission/state';
import { IRewardState } from '_state_mgmt/reward/state';
import { ISurveyState } from '_state_mgmt/survey/state';
import { IUserState } from '_state_mgmt/user/state';

export interface IAction<T = string, P = any> {
  type: T;
  payload: P;
}

export type IReducer<S = any, A = any> = (state: S, action: IAction<A>) => S;

export interface Deps {
  apiService: ApiService;
  appService: AppService;
  crashlyticsService: ICrashlyticsService;
  httpService: IHttpService;
  performanceTrackingService: IPerformanceTrackingService;
  raiseService: RaiseService;
  nativeHelperService: NativeHelperService;
  stateSnapshot: { get: () => IGlobalState; set: (state: IGlobalState) => void };
  logger: Logger;
  eventTrackerService: EventTrackerService;
  ccpaService: CCPAService;
  remoteConfigService: RemoteConfigService;
  authService: Auth0;
}
export type IGlobalState = {
  core: ICoreState;
  user: IUserState;
  game: IGameState;
  mission: IMissionState;
  reward: IRewardState;
  survey: ISurveyState;
  messages: IMessagesState;
  cardLink: ICardLinkState;
  giftCard: IGiftCardState;
};

export interface IGlobalContext {
  state: IGlobalState;
  dispatch: Dispatch<IAction<any, any>>;
  deps: Deps;
}

export enum HTTP_METHOD {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE'
}

export interface ITokenSet {
  accessToken: string;
  refreshToken: string;
  accessTokenExpiryTime: string;
  sywToken?: string;
  /**
   * @deprecated deviceId does not belong to ITokenSet anymore.
   */
  deviceId?: string;
  scope?: string;
  expiresIn?: number;
  tokenType?: string;
}

export enum FeatureFlag {
  API_OVERRIDE = 'API_OVERRIDE',
  TEST_ONLY = 'TEST_ONLY',
  ENV_BANNER = 'ENV_BANNER',
  BUTTON_DEBUG = 'BUTTON_DEBUG',
  LOG_LEVEL_INFO = 'LOG_LEVEL_INFO',
  LOG_LEVEL_DEBUG = 'LOG_LEVEL_DEBUG',
  LOG_LEVEL_WARN = 'LOG_LEVEL_WARN',
  LOG_LEVEL_ERROR = 'LOG_LEVEL_ERROR',
  LOG_VIEW = 'LOG_VIEW',
  STREAK_REDEEM_CARD = 'STREAK_REDEEM_CARD',
  SAILTHRU_TRACKING = 'SAILTHRU_TRACKING',
  SURVEY = 'SURVEY',
  PREDICTIVE_MAP_SEARCH = 'PREDICTIVE_MAP_SEARCH',
  MISSIONS_SEARCH = 'MISSIONS_SEARCH',
  SEE_ALL_TRANSACTIONS = 'SEE_ALL_TRANSACTIONS',
  WALLET_YOUR_GC = 'WALLET_YOUR_GC'
}

export type DateLike = Date | string | number;

export type HTMLString = string;

export interface IGameLevelUI {
  updatedAt: DateLike;
  level: number;
}

export interface ICCPASetting {
  allow: boolean;
}

export interface IOnboarding {
  seenOnboarding: boolean;
}

export enum ButtonCreativeType {
  HERO = 'hero',
  CAROUSEL = 'carousel',
  LIST = 'list',
  GRID = 'grid',
  DETAIL = 'detail',
  OTHER = 'other'
}

export interface ISearchHistoryItem {
  keyword: string;
  lastUpdatedAt: DateLike;
}

export enum TooltipKey {
  EARN = 'earn',
  REWARDS = 'rewards',
  MISSIONS = 'missions',
  STREAK = 'streak'
}

export interface IApiOverrideRule {
  key: string;
  description?: string;
  matcher: string; // reg exp
  status: number;
  response: any;
}

export namespace RemoteConfigModel {
  export interface IAppVersion {
    suggested: number;
    required: number;
  }

  export interface IEarnSection {
    name: string;
    order: number;
    visible: boolean;
    title: string;
    description: string;
    seeAllButton: boolean;
  }
}

export enum RUNTIME_ENV {
  LIVE = 'live',
  DEV_TEST = 'test'
}

export interface IMeasure {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface IExpirePointsBannerDataSet {
  lastPointsDate: DateLike;
  lastAcceptedDate: DateLike;
}

export interface ISailthruMessage {
  createdAt: string;
  custom: {
    [k: string]: string;
  };
  htmlText: string;
  id: string;
  isRead: boolean;
  text: string;
  title: string;
  type: string;
  cardImageUrl?: string;
  share: boolean;
}

export interface ITutorialStatus {
  isBannerWatched: boolean;
  date: string;
}
