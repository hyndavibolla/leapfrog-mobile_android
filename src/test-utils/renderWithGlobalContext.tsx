import React, { ReactElement } from 'react';
import { render } from '@testing-library/react-native';

import { getInitialState, GlobalProvider } from '../state-mgmt/GlobalState';
import { Deps, IGlobalState, IReducer } from '../models/general';
import { getMockDeps } from './getMockDeps';
import { getMockReducer } from './getMockReducer';

export const renderWithGlobalContext = (
  children: ReactElement,
  deps: Deps = getMockDeps(),
  initState: IGlobalState = getInitialState(),
  combinedReducers: IReducer<IGlobalState, any> = getMockReducer()
) => {
  const renderAPI = render(
    <GlobalProvider deps={deps} initState={initState} combinedReducers={combinedReducers}>
      {children}
    </GlobalProvider>
  );
  return { ...renderAPI, mockReducer: combinedReducers, deps };
};
