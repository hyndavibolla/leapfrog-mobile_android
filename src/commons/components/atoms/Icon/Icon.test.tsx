import React from 'react';

import Icon, { Props } from './Icon';
import { Deps } from '_models/general';
import { renderWithGlobalContext } from '_test_utils/renderWithGlobalContext';
import { COLOR, FONT_SIZE, ICON } from '_constants';

describe('Icon', () => {
  let props: Props;
  let deps: Deps;

  beforeEach(() => {
    props = {
      name: ICON.ARROW_RIGHT
    };
  });

  it('should render Default Arrow Right Icon', () => {
    const { toJSON } = renderWithGlobalContext(<Icon {...props} />, deps);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render Arrow Right Icon with Custom Props', () => {
    const nonDefaultProps: Props = {
      ...props,
      color: COLOR.PURPLE,
      size: FONT_SIZE.EXTRA_TINY
    };
    const { toJSON } = renderWithGlobalContext(<Icon {...nonDefaultProps} />, deps);
    expect(toJSON()).toMatchSnapshot();
  });
});
