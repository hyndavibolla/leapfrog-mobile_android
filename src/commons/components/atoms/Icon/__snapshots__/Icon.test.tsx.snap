// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render Arrow Right Icon with Custom Props 1`] = `
<Text
  accessibilityLabel="icon-text"
  accessible={true}
  allowFontScaling={false}
  selectable={false}
  style={
    Array [
      Object {
        "color": "#B361DF",
        "fontSize": 10,
      },
      undefined,
      Object {
        "fontFamily": "sywmax",
        "fontStyle": "normal",
        "fontWeight": "normal",
      },
      Object {},
    ]
  }
  testID="icon-text"
>
  
</Text>
`;

exports[`Icon should render Default Arrow Right Icon 1`] = `
<Text
  accessibilityLabel="icon-text"
  accessible={true}
  allowFontScaling={false}
  selectable={false}
  style={
    Array [
      Object {
        "color": "#0066CC",
        "fontSize": 20,
      },
      undefined,
      Object {
        "fontFamily": "sywmax",
        "fontStyle": "normal",
        "fontWeight": "normal",
      },
      Object {},
    ]
  }
  testID="icon-text"
>
  
</Text>
`;
